### ISSUE TITLE
Memory occupation grows up

### ISSUE THREAD

**Issue by giobus75** [deephealthproject/eddl#212]
Hi, 
I'm trying to train a customized version of a VGG16 with a huge dataset with the python version of my code and the process was killed due to an oom. To understand if the problem was related to python bindings I wrote a piece of code in C++ keeping as simple as possible the implementation to replicate the issue. The code is [here](https://github.com/deephealthproject/promort_pipeline/blob/mem_test/cpp/promort_classification_training.cpp).
In the loop across batches, if I comment rows 137-139 and leave uncommented the 136 one (train_batch) everything works fine. On the contrary, if I comment train_batch row and keep lines 137-139 uncommented, as I do during the evaluation of a validation set, memory occupation keeps increasing.

**Comment by MicheleCancilla** [comment]
Hello @giobus75,

these two lines leak memory:
```cpp
output = getOutput(out); 
cout << output->select({ to_string(0) }) << endl;
```
`Tensor::select` and `getOutput` functions return a new `Tensor*` which should be destroyed.
Try to change them to:
```cpp
output = getOutput(out); 
Tensor* select_tensor = output->select({ to_string(0) });
cout << select_tensor << endl;
delete output;
delete select_tensor;
```

**Comment by giobus75** [comment]
Hi @MicheleCancilla,
I tried your hint ([here the code](https://github.com/deephealthproject/promort_pipeline/blob/mem_test/cpp/promort_classification_training.cpp)) but the problem remains. The memory increases also if I leave only the ```forward``` function call.
I also tried to replace the forward API version with the low-level version: (```net->forward( { x } );```) but the memory keep increasing.

**Comment by RParedesPalacios** [comment]
Hi, 

then if you run:

train_batch(net, { x }, { y }, indices);

instead of:

forward(net, { x });

is then ok??

**Comment by RParedesPalacios** [comment]
I have tested this example that use forward function:

https://github.com/deephealthproject/eddl/blob/master/examples/nn/1_mnist/9_mnist_mlp_func.cpp

and effectively memory grows up! i will check and repair asap

**Comment by giobus75** [comment]
> Hi,
> 
> then if you run:
> 
> train_batch(net, { x }, { y }, indices);
> 
> instead of:
> 
> forward(net, { x });
> 
> is then ok??

Yes, it is

**Comment by giobus75** [comment]
> I have tested this example that use forward function:
> 
> https://github.com/deephealthproject/eddl/blob/master/examples/nn/1_mnist/9_mnist_mlp_func.cpp
> 
> and effectively memory grows up! i will check and repair asap

Thanks a lot, @RParedesPalacios !

**Comment by RParedesPalacios** [comment]
Hi, i found it, is fixed in develop branch, please check

**Comment by giobus75** [comment]
> Hi, i found it, is fixed in develop branch, please check

Hi @RParedesPalacios, no more memory growth. Thank you again.

**Comment by diegobenedicto** [comment]
Just mention that I have experienced this memory leak in Multiple Sclerosis Segmentation training with EDDL 0.7.1, pyEDDL 0.9.0, ECVL 0.2.3, pyECVL 0.5.1
![image](https://user-images.githubusercontent.com/2478595/100996907-f09b4080-3559-11eb-9a1a-e91f2ecd06c0.png)
Good to know it has been solved in EDDL 0.8a
